From 5721b18fc4084a82dbffe4fc41b6e86a17cc8e1c Mon Sep 17 00:00:00 2001
From: HuyLe <hule@amperecomputing.com>
Date: Fri, 12 Aug 2022 09:11:35 +0700
Subject: [PATCH 3/3] add try-catch to avoid get service fail

Currently, the buttons don't work in case the system is a single host,
the software checks whether the platform is multi-host or single host by
checking the dbus object path
"/xyz/openbmc_project/Chassis/Buttons/HostSelector" is valid or not. But
it always returns true.
The root cause comes from the "getService", it did not check whether the
dbus call is successful or not.
This commit adds try-catch to handle the case the dbus call fails, it
will return a null string.

Signed-off-by: HuyLe <hule@amperecomputing.com>
---
 src/button_handler.cpp | 17 +++++++++++------
 1 file changed, 11 insertions(+), 6 deletions(-)

diff --git a/src/button_handler.cpp b/src/button_handler.cpp
index 645602d..b549b42 100644
--- a/src/button_handler.cpp
+++ b/src/button_handler.cpp
@@ -133,12 +133,17 @@ std::string Handler::getService(const std::string& path,
     auto method = bus.new_method_call(mapperService, mapperObjPath, mapperIface,
                                       "GetObject");
     method.append(path, std::vector{interface});
-    auto result = bus.call(method);
-
-    std::map<std::string, std::vector<std::string>> objectData;
-    result.read(objectData);
-
-    return objectData.begin()->first;
+    try
+    {
+        auto result = bus.call(method);
+        std::map<std::string, std::vector<std::string>> objectData;
+        result.read(objectData);
+        return objectData.begin()->first;
+    }
+    catch (const sdbusplus::exception::exception& e)
+    {
+        return std::string();
+    }
 }
 size_t Handler::getHostSelectorValue()
 {
-- 
2.25.1

